src/ucp/api/ucp.h

```diff
@@ -3135,6 +3108,159 @@ ucs_status_ptr_t ucp_worker_flush_nb(ucp_worker_h worker, unsigned flags,
+  UCP_AM_FIELD_IOVEC_SIZE = UCS_BIT(1)
+};
+
+/**
+ * @brief Tuning parameters for the UCP active message.
+ *
+ * The structure defines the parameters that are used for the
+ * UCP active message function
+ */
+typedef struct ucp_am_params {
+  uint64_t field_mask ;
+  uint64_t flags ;
+  /* Number of elements that the ucp library should allocate in the 'iovec'
+   * for the callback to fill in.
+   */
+  size_t iovec_size ;
+} ucp_am_params_t ;
+
+/*
+ * The ucp library will drive a function on the arrival of each fragment
+ * of data from the client.
+ */
+typedef void (*ucp_am_data_function_t)(
+                                       void *target,
+                                       void *source,
+                                       size_t bytes,
+                                       void *cookie
+                                      ) ;
+/*
+ * The ucp library will drive a function when the transfer from the
+ * client is complete.
+ */
+typedef ucs_status_t (*ucp_am_local_function_t)(
+                                                void *arg,
+                                                void *cookie,
+                                                ucp_dt_iov_t *iovec,
+                                                size_t iovec_length
+                                               ) ;
+
+/*
+ * Structure for communicatiob between the active message
+ * callback and the ucp library. The ucp library sets up 'iovec_max_length'
+ * to indicate the length of the iovec in this structure; the remaining
+ * fields are set by the callback to indicate what should happen
+ * with the remaining data from the client. In the initial implementation,
+ * iovec_max_length is always 1.
+ */
+typedef struct ucp_am_recv {
+    /* Function to be driven when data transfer is complete */
+    ucp_am_local_function_t local_fn;
+    /* Argument to be passed to local_fn */
+    void                   *cookie;
+    /* Function to be driven when each fragment of data transfer is complete
+     * In the initial implementation, all data is transferred in one fragment, so
+     * the data_fn si driven once just before the local_fn
+     */
+    ucp_am_data_function_t  data_fn;
+    /* Argument to be passed to data_fn */
+    void                   *data_cookie ;
+    /* Sise of iovec allocated by the ucp library */
+    size_t                  iovec_max_length;
+    /* Size of iovec filled in by the callback */
+    size_t                  iovec_length;
+    /* iovec indicating where the data from the client is to be placed */
+    ucp_dt_iov_t            iovec[1];
+  } ucp_am_recv_t;
+
+  /**
+   * @ingroup UCP_ENDPOINT
+   * @brief Callback to process incoming Active Message.
+   *
+   * When the callback is called, @a flags indicates how @a data should be handled.
+   *
+   * @param [in]  arg      User-defined argument.
+   * @param [in]  data     Points to the received data header.
+   * @param [in]  length   Length of data header.
```

Comment:
What is the header

Reply:
Just to clarify. On the initiator side we don't sent any header. On the receiver side it is explicitly referenced and it is not clear how sender defines one.